add interfaces to allow packet-level inspection for pre/post processing

jdef opened the pull request:
I've been thinking of adding support for "dnstap" to mesos-dns. Having direct access to the DNS data frames would be *much* more efficient (and accurate) than packing/unpacking Msg objects.

This PR defines additional, optional fields to Server that allow callers to optionally inject pre/post processing for DNS data frames.

jdef commented:
xref https://github.com/mesosphere/mesos-dns/pull/212

miekg commented:
Yes, nice. Adding some small comments.

miekg commented on server.go:
```diff
@@ -225,6 +251,10 @@ type Server struct {
 	// If NotifyStartedFunc is set is is called, once the server has started listening.
 	NotifyStartedFunc func()
+	// ReaderBuilder is optional, allows customization of the process that reads DNS frames
+	ReaderBuilder ReaderBuilder
```
Don't really like the 'Builder' part of the name, would ReaderFunc be more appropriate? And the comment needs a closing dot.

miekg commented on server.go:
```diff
@@ -197,6 +198,31 @@ func HandleFunc(pattern string, handler func(ResponseWriter, *Msg)) {
+	io.Writer
+}
+
+type Reader interface {
```
Some docs strings here?

miekg commented:
..and maybe a small Example function that shows a manipulation, like set the message id to zero (or whatever)?

jdef commented:
@miekg I think I've addressed your concerns, let me know if there's anything else you'd like to see in this PR.

I went with DecorateReader/Writer instead of Reader/WriterFunc because the later seems to more closely align with the func-as-an-interface-adapter pattern (like Handler/HandlerFunc) which is definitely not the case here. 'Decorate' seems pretty explicit as to the intent.

miekg commented on server.go:
```diff
@@ -197,6 +198,43 @@ func HandleFunc(pattern string, handler func(ResponseWriter, *Msg)) {
 	DefaultServeMux.HandleFunc(pattern, handler)
 }
 
+// Writer writes DNS data frames; each call to Write should send an entire frame.
```
I would just use DNS message as I use this term throughout the lib. Or maybe (raw) DNS message to distinguish them from the (parsed) DNS message.

miekg commented:
Looks good. Nit about using the word 'frame' which is not a term used in DNS (a lot).

jdef commented:
updated docs, PTAL

tsenart commented on server.go:
```diff
@@ -197,6 +198,43 @@ func HandleFunc(pattern string, handler func(ResponseWriter, *Msg)) {
+
+// DecorateWriter is a decorator hook for extending or supplanting the functionality of a Writer.
+// Implementations should never return a nil Writer.
+type DecorateWriter func(Writer) Writer
```
Nice :-)
